src/App.tsx

```diff
@@ -1,24 +1,190 @@
+import { TodoAppFooter } from './components/TodoAppFooter';
+import { ErrorNotification } from './components/ErrorNotification';
+import { Todo } from './types/Todo';
+import * as todoService from './api/todos';
+import { ErrorMessage } from './utils/errorMessages';
+import { TodoList } from './components/TodoList';
+import { FilterLink } from './types/FilterLinkTypes';
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [selectedFilter, setSelectedFilter] = useState(FilterLink.All);
+  const [errorMessage, setErrorMessage] = useState(ErrorMessage.Default);
+  const [isRequesting, setIsRequesting] = useState(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [processingTodoIds, setProcessingTodoIds] = useState<number[]>([]);
+
+  useEffect(() => {
+    todoService.getTodos(todoService.USER_ID)
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage(ErrorMessage.Load);
+      });
+  }, []);
+
+  const handleAddTodo = (todoTitle: string): Promise<void> => {
+    setTempTodo({
+      id: 0,
+      userId: todoService.USER_ID,
+      title: todoTitle,
+      completed: false,
+    });
+
+    setIsRequesting(true);
+
+    return todoService
+      .addTodo(todoTitle)
+      .then((newTodo) => {
+        setTodos((prevTodos) => [...prevTodos, newTodo]);
+      })
+      .catch(() => {
+        setErrorMessage(ErrorMessage.Add);
+      })
+      .finally(() => {
+        setIsRequesting(false);
+        setTempTodo(null);
+      });
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todoId]);
+
+    todoService
+      .deleteTodo(todoId)
+      .then((() => {
+        setTodos((prevTodos) => prevTodos.filter(todo => todo.id !== todoId));
+      }))
+      .catch(() => {
+        setErrorMessage(ErrorMessage.Delete);
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todoId),
+        );
+      });
+  };
+
+  const handleCompletedChange = (todo: Todo) => {
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todo.id]);
+
+    todoService
+      .updateTodo({
+        ...todo,
+        completed: !todo.completed,
+      })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map(currentTodo => (
+          currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo
+        )));
+      })
+      .catch(() => {
+        setErrorMessage(ErrorMessage.Update);
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todo.id),
+        );
+      });
+  };
+
+  const handleDeleteAllCompletedTodos = () => {
+    todos
+      .filter((todo) => todo.completed)
+      .forEach((todo) => handleDeleteTodo(todo.id));
```

Comment:
It is not necessary to go around the array twice, it is enough to move the filtering condition to the middle of the `forEach` and go through the array once instead of twice